src/libraries/System.Runtime/tests/System/StringTests.cs

```diff
@@ -1838,5 +1838,39 @@ public void TestStringSearchCacheSynchronization(CompareOptions options)
 
             Task.WaitAll(tasks);
         }
+
+        [Fact]
+        public static void EqualityTests_AsciiOptimizations()
+        {
+            for (int i = 0; i < 128; i++)
+            {
+                for (int j = 0; j < 128; j++)
+                {
+                    bool expectedEqualOrdinal = i == j;
+                    bool expectedEqualOrdinalIgnoreCase = (i == j) || ((i | 0x20) >= 'a' && (i | 0x20) <= 'z' && ((i | 0x20) == (j | 0x20)));
+
+                    string s1 = "aaaaaaa" + (char)i + "bbbbbbb";
+                    string s2 = "aaaaaaa" + (char)j + "bbbbbbb";
+
+                    bool actualEqualOrdinal = string.Equals(s1, s2, StringComparison.Ordinal);
+                    bool actualEqualOrdinalIgnoreCase = string.Equals(s1, s2, StringComparison.OrdinalIgnoreCase);
+
+                    int actualCompareToOrdinal = string.Compare(s1, s2, StringComparison.Ordinal);
+                    int actualCompareToOrdinalIgnoreCase = string.Compare(s1, s2, StringComparison.OrdinalIgnoreCase);
+
+                    try
+                    {
+                        Assert.Equal(expectedEqualOrdinal, actualEqualOrdinal);
+                        Assert.Equal(expectedEqualOrdinal, actualCompareToOrdinal == 0);
+                        Assert.Equal(expectedEqualOrdinalIgnoreCase, actualEqualOrdinalIgnoreCase);
+                        Assert.Equal(expectedEqualOrdinalIgnoreCase, actualCompareToOrdinalIgnoreCase == 0);
+                    }
+                    catch (Exception ex)
+                    {
+                        throw new Exception($"Chars U+{i:X4} ('{(char)i}') and U+{j:X4} ('{(char)j}') did not compare as expected. Expected: Ordinal = {expectedEqualOrdinal}, OrdinalIgnoreCase = {expectedEqualOrdinalIgnoreCase}.", ex);
```

Comment:
nit, maybe log the actuals?

Reply:
They're logged as part of the inner exception. The inner exception doesn't contain the values for _i_ and _j_, though, so the outer exception helps provide that context.